File: Irrgarten_Ruby/Irrgarten_Ruby_P4/Player.rb
#encoding:utf-8

require_relative 'Irrgarten'
require_relative 'Weapon'
require_relative 'Shield'
require_relative 'LabyrinthCharacter'


module Irrgarten
	class Player < LabyrinthCharacter
		@@MAX_WEAPONS = 2;
		@@MAX_SHIELDS = 3;
		@@INITIAL_HEALTH = 1;
		@@HITS2LOSE = 3;
		
		def initialize(number, intelligence, strength) #char,float,float
			super("Player ##{@number}",intelligence,strength,@@INITIAL_HEALTH)
			
			@number = number
			@consecutive_hits = 0 
			@weapons = []
			@shields = []
		end
		
		private
		
		def receive_weapon (w) #weapon
			
			@weapons.delete_if do |w| 
				w.discard 
			end
			
			size = @weapons.size
			if (size < @@MAX_WEAPONS)
				@weapons.push(w)
			end
		end
		
		def receive_shield (s) #shield
			 @shields.delete_if do |s| 
        			s.discard
      			end
			
			size = @shields.size
			if (size < @@MAX_SHIELDS)
				@shields.push(s)
			end
		end
		
		def new_weapon ()
			w =  Weapon.new(Dice.weapon_power,Dice.uses_left)
		end
		
		def new_shield ()
			s = Shield.new(Dice.shield_power, Dice.uses_left)
		end
		
		def manage_hit(received_attack)
			defense = defensive_energy
			lose = false
			
			if (defense < received_attack)
				got_wounded
				inc_consecutive_hits
			else
				reset_hits
			end
			
			if (@consecutive_hits == @@HITS2LOSE || dead)
				reset_hits
				lose=true
			else
				lose=false
			end
			
			lose
		end
		
		def reset_hits ()
			@consecutive_hits = 0
		end
		
		def inc_consecutive_hits ()
			@consecutive_hits += 1
		end
		
		protected
		
		attr_reader :weapons, :shields, :consecutive_hits
		
		def sum_weapons ()
			suma = 0.0
			for weapon in @weapons
				suma += weapon.attack
			end
			suma	
		end
		
		def sum_shields ()
			suma = 0.0
			for shield in @shields
				suma += shield.protect
			end
			suma
		end
		
		def defensive_energy()
			suma = @intelligence + sum_shields
		end
		
		public
		
		attr_reader :number
		
		def copiar_player(otro)
			super(otro)
			@number = otro.number
			@consecutive_hits = otro.consecutive_hits
			@weapons = otro.weapons
			@shields = otro.shields
		end
		
		def resurrect() 
			if (dead())
				@health = @@INITIAL_HEALTH
				@weapons.clear()
				@shields.clear()
			end
		end
		
		def move (direction, valid_moves) #Directions, Directions[]
			dir = direction
			
			size = valid_moves.size
			
			contained = valid_moves.include?(direction)
			
			if (!contained && size>0)
				dir = valid_moves[0]
			end
			
			dir
		end	
		
		def attack()
			a = @strength + sum_weapons()
		end
		
		def defend (received_attack) #float
			manage_hit(received_attack)
		end
		
		def receive_reward()
			w_reward = Dice.weapons_reward
			s_reward = Dice.shields_reward
			
			for i in 0...w_reward
				wnew = new_weapon
				receive_weapon(wnew)
			end
			
			for i in 0...s_reward
				snew = new_shield
				receive_shield(snew)
			end
			
			extra_health = Dice.health_reward
			@health += extra_health
		end
		
		def to_s()
			armas = ""
			escudos = ""
			
			@weapons.each do |w|
  				armas += w.to_s + "\n"
			end
			
			@shields.each do |s|
  				escudos += s.to_s + "\n"
			end
			"#{super}, Number: #{@number}, ConsecutiveHits: #{@consecutive_hits} \n Armas: \n #{armas} \n Escudos: \n #{escudos}"
		end
	end
end

File: Irrgarten_Java/Irrgarten_P5/src/irrgarten/Player.java

package irrgarten;

import java.util.ArrayList;

public class Player extends LabyrinthCharacter{
    private static final int MAX_WEAPONS = 2;
    private static final int MAX_SHIELDS = 3;
    private static final int INITIAL_HEALTH = 3;
    private static final int HITS2LOSE = 3;
    private WeaponCardDeck weaponCardDeck;
    private ShieldCardDeck shieldCardDeck;
    
    private char number;
    private int consecutiveHits=0;
    private ArrayList<Weapon> weapons;
    private ArrayList<Shield> shields;
    
    
    private void receiveWeapon( Weapon w){ 
        for(int i=0; i< weapons.size(); i++){
            boolean discard=weapons.get(i).discard();
            
            if (discard){
                weapons.remove(i);
            }
        }
        
        int size= weapons.size();
        if (size< MAX_WEAPONS){ 
            weapons.add(w);
        }
    }
    
    private void receiveShield(Shield s){ 
        for(int i=0; i< shields.size(); i++){
            boolean discard=shields.get(i).discard();
            
            if (discard){
                shields.remove(i);
            }
        }
        
        int size= shields.size();
        if (size< MAX_SHIELDS){ 
            shields.add(s);
        }
    }
    
    private Weapon newWeapon(){
        return weaponCardDeck.nextCard();
    }
    
    private Shield newShield(){
        return shieldCardDeck.nextCard();
    }
    
    protected float sumWeapons(){
        float suma = 0.0f;
        for (int i=0; i< weapons.size(); i++){
            suma += weapons.get(i).attack();
        }
        return suma;
    }
    
    protected float sumShields(){
        float suma = 0.0f;
        for (int i=0; i< shields.size(); i++){
            suma += shields.get(i).protect();
        }
        return suma;
    }
    
    protected float defensiveEnergy(){
        return super.getIntelligence() + sumShields();
    }
    
    private boolean manageHit(float receivedAttack){
        float defense= defensiveEnergy();
        boolean lose=false;
        
        if(defense<receivedAttack){
            super.getWounded();
            incConsecutiveHits();
        } else {
            resetHits();
        }

        if(consecutiveHits==HITS2LOSE || (super.dead())) {
            resetHits();
            lose=true;
        } else {
            lose=false;
        }
        return lose;
    }
    
    private void resetHits(){
        consecutiveHits=0;
    }
    
   
    private void incConsecutiveHits(){
        consecutiveHits++;
    }
    
    public Player ( char number , float intelligence, float strength){
           super("Player"+number,intelligence,strength,INITIAL_HEALTH);
           this.number=number;
           this.weapons = new ArrayList<>();
           this.shields = new ArrayList<>();
           this.weaponCardDeck = new WeaponCardDeck();
           this.shieldCardDeck = new ShieldCardDeck();
    }

    public Player(Player other){
        super(other);
        this.number=other.number;
        this.consecutiveHits=other.consecutiveHits;
        this.weapons = other.weapons;
        this.shields = other.shields;
        this.weaponCardDeck = other.weaponCardDeck;
        this.shieldCardDeck =other.shieldCardDeck;
    }
    
    public void resurrect(){
        assert super.dead() : "Estas intentando resucitar sin estar muerto";
       
        super.setHealth(INITIAL_HEALTH);
        this.weapons.clear();
        this.shields.clear();
    }
    
    
    
    public char getNumber(){
        return this.number;
    }
    
    public Directions move ( Directions direction, ArrayList<Directions> validMoves){ 
        Directions dir=direction;
        
        int size= validMoves.size();
       
        boolean contained=validMoves.contains(direction); 
        
        if (!contained && size>0){
            dir=validMoves.get(0);
        }
        
          return dir;
    }
    
    @Override
    public float attack(){
        return super.getStrength()+ sumWeapons();
    }
    
    @Override
    public boolean defend(float receivedAttack){
        return manageHit(receivedAttack);
    }
    
    public void receiveReward(){
        int wReward= Dice.weaponsReward();
        int sReward= Dice.shieldsReward();
        
        for (int i=0; i< wReward; i++){
            Weapon wnew= newWeapon();
            this.receiveWeapon(wnew);
        }
        for (int i=0; i< sReward; i++){
            Shield snew= newShield();
            this.receiveShield(snew);
        }
        int extraHealth= Dice.healthReward();
        super.setHealth((super.getHealth()+extraHealth));
    }
    
    @Override
    public String toString(){
        return (super.toString()+ ", Number: " + number 
        + ", ConsecutiveHits: " + consecutiveHits + " \nArmas: \n" + weapons + "\nEscudos: \n" + shields + "\nBaraja Armas: \n" + weaponCardDeck + "\nBaraja Escudos: \n" + shieldCardDeck + "\n");
    }
    
    
    
}
